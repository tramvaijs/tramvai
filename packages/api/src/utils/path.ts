import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
