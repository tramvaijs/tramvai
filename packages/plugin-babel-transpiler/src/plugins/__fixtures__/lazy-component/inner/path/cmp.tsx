import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
