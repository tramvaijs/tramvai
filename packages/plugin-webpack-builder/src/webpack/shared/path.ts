import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
