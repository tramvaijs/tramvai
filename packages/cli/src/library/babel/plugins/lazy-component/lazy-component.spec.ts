import path from 'path';
import pluginTester from 'babel-plugin-tester';
import babelConfig from '../../index';

pluginTester({
  plugin: {}, // плагин уже есть в основном конфиге, если подключить ещё тут то babel упадёт
  pluginName: 'lazy-component',
  filename: path.join(__dirname, '..', '__fixtures__', 'lazy-component', 'test'),
  babelOptions: babelConfig({
    typescript: true,
    generateDataQaTag: false,
    loader: false,
    modern: true,
    hot: true,
  }),
  tests: {
    'not lazy': {
      fixture: 'not-lazy.ts',
      snapshot: true,
    },
    'tramvai react but not lazy': {
      fixture: 'not-lazy-import.ts',
      snapshot: true,
    },
    base: {
      fixture: 'base.ts',
      snapshot: true,
    },
    'long import': {
      fixture: 'long-import.ts',
      snapshot: true,
    },
    dynamic: {
      fixture: 'dynamic.ts',
      snapshot: true,
    },
    comments: {
      fixture: 'comments.ts',
      snapshot: true,
    },
    hmr: {
      fixture: 'lazy+hmr.ts',
      snapshot: true,
    },
  },
});
