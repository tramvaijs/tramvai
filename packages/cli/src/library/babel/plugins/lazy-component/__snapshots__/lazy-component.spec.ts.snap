// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy-component base: base 1`] = `

import { lazy } from '@tramvai/react';

lazy(() => import('./cmp'))

      ↓ ↓ ↓ ↓ ↓ ↓

import { lazy } from '@tramvai/react';
lazy({
  chunkName() {
    return 'cmp';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () => import(/* webpackChunkName: "cmp" */ './cmp'),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./cmp');
  },
});


`;

exports[`lazy-component comments: comments 1`] = `

import { lazy } from '@tramvai/react';

lazy(() => import(/* webpackChunkName: 'main' */'./inner/first'))
lazy(() => import(/* stupid comments */'./inner/second'))
lazy(() => import(/* webpackPreload: true */'./inner/third'))

lazy(() => import(/* webpackPrefetch: true, webpackChunkName: "component"  */'./cmp'))

      ↓ ↓ ↓ ↓ ↓ ↓

import { lazy } from '@tramvai/react';
lazy({
  chunkName() {
    return 'main';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () => import(/* webpackChunkName: "main" */ './inner/first'),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./inner/first');
  },
});
lazy({
  chunkName() {
    return 'inner-second';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () =>
    import(
      /* webpackChunkName: "inner-second" */
      /* stupid comments */
      './inner/second'
    ),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./inner/second');
  },
});
lazy({
  chunkName() {
    return 'inner-third';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () =>
    import(
      /* webpackChunkName: "inner-third" */
      /* webpackPreload: true */
      './inner/third'
    ),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./inner/third');
  },
});
lazy({
  chunkName() {
    return 'component';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "component" */ './cmp'),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./cmp');
  },
});


`;

exports[`lazy-component dynamic: dynamic 1`] = `

import { lazy } from '@tramvai/react';

lazy((props) => import(\`./inner/\${props.name}.js\`))

      ↓ ↓ ↓ ↓ ↓ ↓

import { lazy } from '@tramvai/react';
lazy({
  chunkName(props) {
    return \`inner-\${props.name}-js\`.replace(/[^a-zA-Z0-9_!§$()=\\-^°]+/g, '-');
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: (props) =>
    import(
      /* webpackChunkName: "inner-[request]" */ \`./inner/\${props.name}.js\`
    ),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve(props) {
    return require.resolveWeak(\`./inner/\${props.name}.js\`);
  },
});


`;

exports[`lazy-component hmr: hmr 1`] = `

import { lazy } from '@tramvai/react';

const MainPage = lazy(() => import('./inner/first'));

      ↓ ↓ ↓ ↓ ↓ ↓

import { lazy } from '@tramvai/react';
const MainPage = lazy(
  (_c = {
    chunkName() {
      return 'inner-first';
    },
    requireSync(props) {
      return __webpack_require__(this.resolve(props));
    },
    isReady(props) {
      const key = this.resolve(props);
      return !!__webpack_modules__[key];
    },
    importAsync: () =>
      import(/* webpackChunkName: "inner-first" */ './inner/first'),
    requireAsync(props) {
      const key = this.resolve(props);
      return this.importAsync(props).then((resolved) => {
        return resolved;
      });
    },
    resolve() {
      return require.resolveWeak('./inner/first');
    },
  })
);
_c2 = MainPage;
var _c, _c2;
$RefreshReg$(_c, 'MainPage$lazy');
$RefreshReg$(_c2, 'MainPage');


`;

exports[`lazy-component long import: long import 1`] = `

import { lazy } from '@tramvai/react';

lazy(() => import('./inner/path/cmp'))

      ↓ ↓ ↓ ↓ ↓ ↓

import { lazy } from '@tramvai/react';
lazy({
  chunkName() {
    return 'inner-path-cmp';
  },
  requireSync(props) {
    return __webpack_require__(this.resolve(props));
  },
  isReady(props) {
    const key = this.resolve(props);
    return !!__webpack_modules__[key];
  },
  importAsync: () =>
    import(/* webpackChunkName: "inner-path-cmp" */ './inner/path/cmp'),
  requireAsync(props) {
    const key = this.resolve(props);
    return this.importAsync(props).then((resolved) => {
      return resolved;
    });
  },
  resolve() {
    return require.resolveWeak('./inner/path/cmp');
  },
});


`;

exports[`lazy-component not lazy: not lazy 1`] = `

const lazy = (imp) => console.log(imp);
const load = lazy(() => import('./cmp'));

(async () => {
  const {default: cmp} = await import('./cmp');

  console.log((cmp as any).actions)
})();

      ↓ ↓ ↓ ↓ ↓ ↓

const lazy = (imp) => console.log(imp);
const load = lazy(() => import('./cmp'));
(async () => {
  const { default: cmp } = await import('./cmp');
  console.log(cmp.actions);
})();


`;

exports[`lazy-component tramvai react but not lazy: tramvai react but not lazy 1`] = `

import { withError } from '@tramvai/react';

export const compose = () => {
  return (cmp) => {
    withError()(cmp);
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { withError } from '@tramvai/react';
export const compose = () => {
  return (cmp) => {
    withError()(cmp);
  };
};


`;
